### PR TITLE
changes based on upstream comments

### PR DESCRIPTION
**Pull request by rolandvarga**

### CODE REVIEW

**Review by vincentcreusotfx — APPROVED** [review]

**Review by louisfelix — CHANGES_REQUESTED** [review]
Good improvement! But I see two problems
1) `resourceList` must be reset on every `Collect` invocation, because if a tag change, we want this new info as soon as it changed
2) I think `resourceList[f.ID]` logic should be applied to resource group config (not only target and tag config)

File: main.go [hunk 1]
```diff
@@ -105,12 +107,15 @@ func (c *Collector) extractMetrics(ch chan<- prometheus.Metric, rm resourceMeta,
 		}
 	}
 
-	infoLabels := CreateAllResourceLabelsFrom(rm)
-	ch <- prometheus.MustNewConstMetric(
-		prometheus.NewDesc("azure_resource_info", "Azure information available for resource", nil, infoLabels),
-		prometheus.UntypedValue,
-		0,
-	)
+	if _, ok := publishedResource[rm.resource.ID]; !ok {
```
Comment (vincentcreusotfx) [inline main.go:110]: I think the best way to do that is not with a bool but more a empty struct and you use the first return as the ok here

File: main.go [hunk 2]
```diff
@@ -148,27 +153,33 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
+	publishedResource = map[string]bool{}
```
Comment (vincentcreusotfx) [inline main.go:156]: should use make(map[string]struct{})

File: main.go [hunk 3]
```diff
@@ -148,27 +153,33 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
...
+		rm.aggregations = filterAggregations(target.Aggregations)
+		rm.resourceURL = resourceURLFrom(target.Resource, rm.metrics, rm.aggregations)
 
-		var err error
-		resource.Resource, err = ac.lookupResourceByID(target.Resource)
-		if err != nil {
-			log.Printf("failed to get resource information for target %s: %v", target.Resource, err)
-			ch <- prometheus.NewInvalidMetric(azureErrorDesc, err)
-			continue
+		if _, ok := resourceList[target.Resource]; !ok {
```
Comment (vincentcreusotfx) [inline main.go:172]: Same idea

File: main.go [hunk 4]
```diff
@@ -212,19 +224,22 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
...
-			if err != nil {
-				log.Printf("failed to get resource information for target %s: %v", f.ID, err)
-				ch <- prometheus.NewInvalidMetric(azureErrorDesc, err)
-				continue
+			var rm resourceMeta
+			rm.metrics = metricsStr
+			rm.aggregations = filterAggregations(resourceTag.Aggregations)
+			rm.resourceURL = resourceURLFrom(f.ID, rm.metrics, rm.aggregations)
+
+			if _, ok := resourceList[f.ID]; !ok {
```
Comment (vincentcreusotfx) [inline main.go:232]: Same ideap

File: main.go [hunk 5]
```diff
@@ -27,6 +27,8 @@ var (
 	invalidMetricChars    = regexp.MustCompile("[^a-zA-Z0-9_:]")
 	azureErrorDesc        = prometheus.NewDesc("azure_error", "Error collecting metrics", nil, nil)
 	batchSize             = 20
+	publishedResource     = map[string]bool{}
```
Comment (vincentcreusotfx) [inline main.go:30]: use a struct{} like in filterTypesInResourceList

File: utils.go [hunk 6]
```diff
@@ -16,7 +16,8 @@ var (
 	resourceNamePosition    = 8
 	subResourceNamePosition = 10
 
-	invalidLabelChars = regexp.MustCompile("[^\\w]")
+	invalidLabelPrefix = regexp.MustCompile(`^[^a-zA-Z_]*`)
```
Comment (vincentcreusotfx) [inline utils.go:19]: is the star ok here ? it means 0 or more right ?

**Review by louisfelix — APPROVED** [review]

File: utils.go [hunk 7]
```diff
@@ -55,11 +56,16 @@ func CreateResourceLabels(resourceID string) map[string]string {
...
-	split := strings.Split(rm.ResourceURL, "/")
+	split := strings.Split(rm.resourceURL, "/")
 	labels["resource_group"] = split[resourceGroupPosition]
 
-	for k, v := range rm.Resource.Tags {
+	for k, v := range rm.resource.Tags {
 		k = strings.ToLower(k)
+
+		if !invalidLabelPrefix.MatchString(k) {
+			k = "_" + k
```
Comment (vincentcreusotfx) [inline utils.go:66]: invalidLabelPrefix contains _ and then we add one after ? not sure I understand then should'nt the regex be named validPrefix ?

**Pull request merged**